### ISSUE TITLE
Suggestion, add info about Google colab python notebook

### ISSUE THREAD

**Issue by 5525345551** [gcp/leela-zero#1223]
leela-chess has a python notebook for Google colab to get use of a K80 GPU for free.

I've made a version for leela-zero:
https://drive.google.com/file/d/1a-Dq_I6NBU11jO-0hOGA9olNV8gMgOPT/view?usp=sharing

@bood also has made a notebook, link in the reddit cbaduk page.
https://www.reddit.com/r/cbaduk/comments/8d9t7i/how_do_i_use_leela_with_google_colab/

Perhaps info can be added to the README.md linking to a wiki page on this project ala leela-chess
https://github.com/glinscott/leela-chess/wiki/Run-Leela-Chess-Zero-client-on-a-Tesla-K80-GPU-for-free-(Google-Colaboratory)

**Comment by l1t1** [comment]
powerful gpu

**Comment by gcp** [comment]
It should be possible to make a COLAB.md file or so including the markup, if you want to redit/rewrite the Leela Chess one for Leeela Zero.

**Comment by hydrogenpi** [comment]
@has someone created a youtube tutorial on how to do this step by step for the general public interested in helping out? I think Google policy says this is okay as long as each person only has one going on at any one time. Chess guys got in trouble because they tried to do multiple instances per person.

**Comment by Alderi-Tokori** [comment]
Note that every person shouldn't run more than one instance of the collab Notebook at the same time, I've heard the LCZ people got a bit of heat from google  because they were running too many instances.

**Comment by bldght** [comment]
Very good, I have trained a game。
[http://zero.sjeng.org/view/201a8c99c5c00af9e88db32efc2d535852e0c8e2e7e94878c737897e52e0fa0f?viewer=wgo](http://zero.sjeng.org/view/201a8c99c5c00af9e88db32efc2d535852e0c8e2e7e94878c737897e52e0fa0f?viewer=wgo)

Game data 201a8c99c5c00af9e88db32efc2d535852e0c8e2e7e94878c737897e52e0fa0f stored in database
1 game(s) (1 self played and 0 matches) played in 17 minutes = 1074 seconds/game, 4649 ms/move, last game took 1073 seconds.

It seems OK.Thanks to @5525345551  and Google.

**Comment by hydrogenpi** [comment]
http://archive.is/aq1iS

**Comment by 3kakakhel** [comment]
I tried this method out because I thought I would be a good learning opportunity for me. I know I have a lot to learn, but I tried it anyway thinking it would work better than the 5 year old gpu I have been using.

I got the notebook onto my google drive, into colab, and running. The only thing is that I endup with the following ouput and I am uncertain if it is actually sending games to the server. All of the games are like the following snippet. Notice the following concerns:

- leelaz unexpectedly fails
- gzip says that {hash}.sgf: no such file or directory.
- Upload fails with curl giving error code 26.
- Network connection fails.

It tries a few times, but I am not sure if it finally goes through or not.

	{
	    "cmd": "selfplay",
	    "hash": "e8601c38352330a313b99d09792649504767901529410cfbddb1d0f102055652",
	    "leelaz_version": "0.13",
	    "options": {
	        "noise": "true",
	        "playouts": "0",
	        "randomcnt": "30",
	        "resignation_percent": "5",
	        "visits": "3201"
	    },
	    "options_hash": "cb9add",
	    "random_seed": "18401068929008911075",
	    "required_client_version": "15"
	}

	Got new job: selfplay
	net: e8601c38352330a313b99d09792649504767901529410cfbddb1d0f102055652.
	Engine has started.
	Infinite thinking time set.
	1 (B D17) 2 (W R4) 3 (B Q17) 4 (W D3) 5 (B C5) 6 (W F4) 7 (B P3) 8 (W Q6) 9 (B R14) 10 (W C15) 11 (B K5) 12 (W E16) 13 (B E17) 14 (W F16) 15 (B G17) 16 (W C8) 17 (B G16) 18 (W G15) 19 (B A17) 20 (W H15) 21 (B C13) 22 (W C17) 23 (B C18) 24 (W F17) 25 (B F18) 26 (W G18) 27 (B E18) 28 (W H17) 29 (B D16) 30 (W M3) 31 (B E15) 32 (W H16) 33 (B E13) 34 (W M5) 35 (B D7) 36 (W D8) 37 (B E8) 38 (W E7) 39 (B C7) 40 (W E6) 41 (B E9) 42 (W B7) 43 (B B6) 44 (W B8) 45 (B C3) 46 (W C2) 47 (B C4) 48 (W C11) 49 (B D4) 50 (W D2) 51 (B E4) 52 (W F3) 53 (B F5) 54 (W G5) 55 (B G6) 56 (W E5) 57 (B F6) 58 (W F7) 59 (B H6) 60 (W G8) 61 (B B2) 62 (W B1) 63 (B A2) 64 (W J4) 65 (B E11) 66 (W B12) 67 (B B13) 68 (W B10) 69 (B J8) 70 (W H9) 71 (B J9) 72 (W H10) 73 (B H8) 74 (W G9) 75 (B J10) 76 (W H11) 77 (B L3) 78 (W A4) 79 (B A5) 80 (W B4) 81 (B H2) 82 (W H3) 83 (B G2) 84 (W F2) 85 (B K3) 86 (W J5) 87 (B G3) 88 (W G4) 89 (B J6) 90 (W K4) 91 (B L4) 92 (W L5) 93 (B K6) 94 (W J3) 95 (B J2) 96 (W H5) 97 (B F1) 98 (W M2) 99 (B M4) 100 (W N4) 101 (B L2) 102 (W L1) 103 (B K2) 104 (W N3) 105 (B N5) 106 (W M6) 107 (B O4) 108 (W O2) 109 (B P2) 110 (W K1) 111 (B resign) Game has ended.
	Score: W+Resign 
	Winner: white
	*ERROR*: The 'leelaz' process died unexpected.
	Uploading game: dd442e9b5b0f47978fe9559e45a0f9c9.sgf for network e8601c38352330a313b99d09792649504767901529410cfbddb1d0f102055652
	gzip: dd442e9b5b0f47978fe9559e45a0f9c9.sgf: No such file or directory
	Upload failed. Curl Exit code: 26
	Network connection to server failed.
	NetworkException: Curl returned non-zero exit code 26
	Retrying in 30 s.
	Upload failed. Curl Exit code: 26
	Network connection to server failed.
	NetworkException: Curl returned non-zero exit code 26
	Retrying in 45 s.
	Upload failed. Curl Exit code: 26
	Network connection to server failed.
	NetworkException: Curl returned non-zero exit code 26
	Retrying in 67 s.
	1 game(s) (1 self played and 0 matches) played in 11 minutes = 700 seconds/game, 6306 ms/move, last game took 554 seconds.

**Comment by 5525345551** [comment]
Looks like the game did not upload because of 'No such file'.
No problems here with it working. 
If it fails try 'Reconnect' and 'Run All'.

**Comment by 3kakakhel** [comment]
Thank you. Do I have to run all the items again or really just the last one?

**Comment by fell111** [comment]
Just tried the COLAB. Here is the result

1 game(s) (1 self played and 0 matches) played in 21 minutes = 1303 seconds/game, 5030 ms/move, last game took 1301 seconds.

It's faster than my desktop! I'll keep it running.

**Comment by Cabu** [comment]
The notebook worked perfectly for me !

**Comment by 3kakakhel** [comment]
I tried it again over the weekend. It seemed to get stuck earlier; when autogtp tries to get info from the server. See the output below.

	AutoGTP v15
	Using 1 thread(s) for GPU(s).
	Starting tuning process, please wait...
	Network connection to server failed.
	NetworkException: Curl returned non-zero exit code 23
	Retrying in 30 s.

I just got the idea that maybe something changed in the notebook or something. I think I must have used bood's notebook because I tried 5525345551's notebook today and it seems to be running. 😃 

The difference between the two is that bood's version tries to use a precompiled version from the user's google drive while 5525345551's builds everything from scratch assuming a generic ubuntu/debian environment?

**Comment by 5525345551** [comment]
@3kakakhel yes you are right, the version I am using is the same as Leela Chess Zero notebook.
It builds everything from scratch when a 'Run All' is done.
Here is a direct link to the notebook : 
[https://colab.research.google.com/drive/1WQfPOFhkahIJSxdIjeSQqK4q30j3T1qF](url)

**Comment by hydrogenpi** [comment]
Amazon AWS offers a free tier for windows and Linux and its 2x vCPU burst but sustained its roughly only 1/2 CPU average. I was thinking that low PO self games could help with diversity and maybe even help with ladder learning since it has to make mistakes to learn from them properly. 

If it was possible to allow very low playout self playing games as part of the contribution, then potentially there can be a lot more clients using the free tier of AWS EC2 using just the cpu only version of leelaz to train.

**Comment by Cabu** [comment]
I never get my notebook working for more than 4 hours :( Each time I need to restart it after 1 to 4 hours :(

**Comment by 5525345551** [comment]
Yeah, i have the same problem. I have it on an old phone that stays awake when charging. Sometimes I get 30+ games and other times only a few. Still better than nothing i suppose.

**Comment by 3kakakhel** [comment]
> Perhaps it is an idea to create your own copy of the notebook to your google drive.

I follow this method, but I still have the same experience as Cabu.

**Comment by fell111** [comment]
It's strange that the program stopped every 3 games. I need to restart the notebook to continue self play. Next time I'll build up my own notebook to try.

**Comment by rexl2018** [comment]
Please take note that don't login a few accounts and run several instances at the same time.
We have many many accounts blocked these days because of multi-instance.

**Comment by WYVYW** [comment]
I started colab after the release of leelaz v. 0.15 but version v 0.13 seems still to be used.
![image](https://user-images.githubusercontent.com/38362247/39841080-ea683d50-53e1-11e8-8447-603d20ca074c.png)
